src/estimators/same_key_aggregator.py

```diff
@@ -94,54 +98,107 @@ def assert_compatible(self, other):
+      reach_noiser_kwargs: a dictionary of other kwargs to be specified in
+        reach_noiser_class. For example, it can be an empty dictionary, or
+        {'random_state': <a np.random.RandomState>}.
+      frequency_noiser_kwargs: a dictionary of other kwargs to be specified in
+        frequency_noiser_class. For example, it can be an empty dictionary, or
+        {'random_state': <a np.random.RandomState>}.
     """
     self.max_freq = max_freq
-    self.one_plus_reach_noiser = None
-    self.histogram_noiser = None
-    if noiser_class is not None:
-      assert epsilon_split > 0 and epsilon_split < 1, (
-          'In StandardizedHistogramEstimator, epsilon_split must be >0 and <1.')
-      # self.one_plus_reach_noiser is the noiser on the estimate of 1+ reach.
-      self.one_plus_reach_noiser = noiser_class(epsilon=epsilon * epsilon_split)
-      # self.histogram_noiser is the noiser on the frequency histogram among
-      # effective keys
-      # For simplicity, suppose the two noisers share the same noiser_class.
-      self.histogram_noiser = noiser_class(
-          epsilon=epsilon * (1 - epsilon_split))
+    self.one_plus_reach_noiser, self.reach_epsilon, self.reach_delta = (
+        StandardizedHistogramEstimator.define_noiser(
+            reach_noiser_class, reach_epsilon, reach_delta, reach_noiser_kwargs))
+    self.histogram_noiser, self.frequency_epsilon, self.frequency_delta = (
+        StandardizedHistogramEstimator.define_noiser(
+            frequency_noiser_class, frequency_epsilon,
+            frequency_delta, frequency_noiser_kwargs))
+    self.total_epsilon, self.total_delta = self.calculate_total_privacy_budget()
 
   def __call__(self, sketch_list):
     ska = StandardizedHistogramEstimator.merge_sketch_list(sketch_list)
     return self.estimate_cardinality(ska)
 
+  @classmethod
+  def define_noiser(cls, noiser_class, epsilon, delta=0, kwargs={}):
+    """A util for defining reach_noiser and frequency_noiser.
+
+    Args:
+      noiser_class: given name of noiser_class.
+      epsilon: given epsilon for the noiser.
+      delta: given delta for the noiser.
+      kwargs: other args to be specified in the noiser_class.
+
+    Returns:
+      a tuple of (<obtained noiser>, <actual epsilon the noiser guarantees>,
+      <actual delta the noiser guarantees>)
+    """
+    if noiser_class is None:
+      return (None, np.Inf, 0)
+    if delta in inspect.getfullargspec(noiser_class)[0]:
+      # If delta is an argument of noiser_class
+      return (noiser_class(epsilon=epsilon, delta=delta, **kwargs),
+              epsilon, delta)
+    # Otherwise, delta is not an argument of noiser_class
+    warnings.warn('The given delta is ignored, since it is not in the noiser.')
+    return (noiser_class(epsilon=epsilon, **kwargs), epsilon, 0)
+
+  def calculate_total_privacy_budget(self):
+    return (self.reach_epsilon + self.frequency_epsilon * self.max_freq,
```

Comment:
This code seems to suggest that the parallel composition theorem cannot be applied to the different buckets in the frequency histogram.  Is this true?  To me, it seems that if one user is added or removed, that user can only affect a single frequency bucket.  Am I wrong about this?

Reply:
Quick answer to this comment (will answer other comments later):
The simple parallel composition assumes delta = 0. It does not work for Gaussian noiser which must have delta > 0. @pasin30055 Is there an efficient parallel composition for delta > 0?

Perhaps, we can use Discrete Laplace for active_key_histogram (to have parallel composition) and Discrete Gaussian for one_plus_reach (since Gaussian is efficient for composition of large # of queries).

Reply:
Sorry for the complexity here. It should be easy to just derive the correct parallel composition for the case with delta. Let me do it over the weekend and hopefully through the math we'll see how it is different from the simple parallel composition.

Reply:
A couple of comments:
- I don't see why parallel composition wouldn't work for GaussianMechanism. Can you please clarify?
- I'd actually support letting the histogram noiser itself deals with its own epsilon, delta. I.e. let's pass in the epsilon, delta and let the noiser computes the noise to be added to each bucket to achieve the input epsilon, delta. This way we can remove the "* self.max_freq" here and possibly in the future support other more complicated noisers. Wdyt?